File: server/java/render/src/main/java/com/metaobjects/render/recover/FieldSpec.java
package com.metaobjects.render.recover;

import java.util.List;
import java.util.Map;

/**
 * One field's recover descriptor. enumValues/enumAlias non-null only for ENUM;
 * min/max non-null only for numeric range constraints; nested non-null only for OBJECT.
 */
public record FieldSpec(
        String name,
        FieldKind kind,
        boolean required,
        boolean array,
        List<String> enumValues,
        Map<String, String> enumAlias,
        Double min,
        Double max,
        RecoverSchema nested) {

    public static FieldSpec scalar(String name, FieldKind kind, boolean required) {
        return new FieldSpec(name, kind, required, false, null, null, null, null, null);
    }

    public static FieldSpec enumField(String name, boolean required,
                                      List<String> values, Map<String, String> aliases) {
        return new FieldSpec(name, FieldKind.ENUM, required, false,
                values == null ? null : List.copyOf(values),
                aliases == null ? Map.of() : Map.copyOf(aliases),
                null, null, null);
    }

    public static FieldSpec range(String name, FieldKind kind, boolean required,
                                  Double min, Double max) {
        return new FieldSpec(name, kind, required, false, null, null, min, max, null);
    }

    public static FieldSpec object(String name, boolean required, boolean array, RecoverSchema nested) {
        return new FieldSpec(name, FieldKind.OBJECT, required, array, null, null, null, null, nested);
    }
}

File: server/csharp/MetaObjects.Render/Recover/FieldSpec.cs
namespace MetaObjects.Render.Recover;

/// <summary>
/// One field's recover descriptor.
/// <see cref="EnumValues"/> and <see cref="EnumAlias"/> are non-null only for <see cref="FieldKind.Enum"/>.
/// <see cref="Min"/> / <see cref="Max"/> are non-null only for numeric range constraints.
/// <see cref="Nested"/> is non-null only for <see cref="FieldKind.Object"/>.
/// </summary>
public sealed record FieldSpec(
    string Name,
    FieldKind Kind,
    bool Required,
    bool Array,
    IReadOnlyList<string>? EnumValues,
    IReadOnlyDictionary<string, string>? EnumAlias,
    double? Min,
    double? Max,
    RecoverSchema? Nested)
{
    /// <summary>Build a plain scalar field (string / int / long / double / boolean).</summary>
    public static FieldSpec Scalar(string name, FieldKind kind, bool required) =>
        new(name, kind, required, false, null, null, null, null, null);

    /// <summary>Build an enum field with its allowed values and optional case/alias map.</summary>
    public static FieldSpec EnumField(
        string name,
        bool required,
        IReadOnlyList<string>? values,
        IReadOnlyDictionary<string, string>? aliases) =>
        new(name, FieldKind.Enum, required, false,
            values == null ? null : new List<string>(values).AsReadOnly(),
            aliases == null
                ? new Dictionary<string, string>()
                : new Dictionary<string, string>(aliases),
            null, null, null);

    /// <summary>Build a numeric field with optional min/max range constraints.</summary>
    public static FieldSpec Range(string name, FieldKind kind, bool required, double? min, double? max) =>
        new(name, kind, required, false, null, null, min, max, null);

    /// <summary>Build a nested object field backed by a child <see cref="RecoverSchema"/>.</summary>
    public static FieldSpec Object(string name, bool required, bool array, RecoverSchema nested) =>
        new(name, FieldKind.Object, required, array, null, null, null, null, nested);
}
